File: android/src/main/kotlin/com/example/image_shape_cropper/ImageShapeCropperPlugin.kt
package com.example.image_shape_cropper

import android.graphics.*
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.ByteArrayOutputStream
import android.util.Log

/** ImageShapeCropperPlugin */
class ImageShapeCropperPlugin: FlutterPlugin, MethodChannel.MethodCallHandler {
  private lateinit var channel: MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "image_shape_cropper")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
    when (call.method) {
      "cropOval" -> {
        val sourcePath = call.argument<String>("sourcePath")
        val angle = call.argument<Double>("angle") ?: 0.0
        val width = call.argument<Double>("width")
        val height = call.argument<Double>("height")
        val scale = call.argument<Double>("scale") ?: 1.0
        val compressFormat = call.argument<String>("compressFormat") ?: "png"
        val compressQuality = call.argument<Int>("compressQuality") ?: 100

        Log.d(
          "ImageShapeCropper",
          "cropOval called with sourcePath: $sourcePath, angle: $angle, width: $width, height: $height, scale: $scale, compressFormat: $compressFormat, compressQuality: $compressQuality"
        )

        if (sourcePath == null) {
          result.error("INVALID_ARGUMENT", "sourcePath is null", null)
          return
        }

        try {
          val bitmap = BitmapFactory.decodeFile(sourcePath)
          if (bitmap == null) {
            Log.e("ImageShapeCropper", "Cannot decode image from path: $sourcePath")
            result.error("INVALID_IMAGE", "Cannot decode image", null)
            return
          }

          // Rotate the bitmap
          val rotatedBitmap = rotateBitmap(bitmap, angle.toFloat())
          Log.d("ImageShapeCropper", "Bitmap rotated by $angle degrees.")

          // Crop to oval with scaling
          val ovalBitmap = cropOval(rotatedBitmap, width, height, scale)
          Log.d("ImageShapeCropper", "Bitmap cropped to oval.")

          // Compress the bitmap
          val byteArrayOutputStream = ByteArrayOutputStream()
          val format = if (compressFormat.equals("png", ignoreCase = true)) {
            Bitmap.CompressFormat.PNG
          } else {
            Bitmap.CompressFormat.JPEG
          }
          ovalBitmap.compress(format, compressQuality, byteArrayOutputStream)
          val bytes = byteArrayOutputStream.toByteArray()

          Log.d("ImageShapeCropper", "Bitmap compressed. Size: ${bytes.size} bytes.")

          result.success(bytes)
        } catch (e: Exception) {
          Log.e("ImageShapeCropper", "Error cropping image: ${e.message}")
          result.error("CROP_FAILED", e.message, null)
        }
      }

      else -> {
        result.notImplemented()
      }
    }
  }

  private fun rotateBitmap(source: Bitmap, angle: Float): Bitmap {
    val matrix = Matrix()
    matrix.postRotate(angle)
    return Bitmap.createBitmap(source, 0, 0, source.width, source.height, matrix, true)
  }

  private fun cropOval(source: Bitmap, desiredWidth: Double?, desiredHeight: Double?, scale: Double): Bitmap {
    val width = desiredWidth ?: source.width.toDouble()
    val height = desiredHeight ?: source.height.toDouble()

    // Ensure scale is at least 1.0
    val effectiveScale = when {
      scale < 1.0 -> 1.0
      scale > 5.0 -> 5.0
      else -> scale
    }

    // Calculate the size of the crop rectangle based on scale
    val cropWidth = (width / effectiveScale).toFloat()
    val cropHeight = (height / effectiveScale).toFloat()

    // Calculate the top-left coordinates for center-cropping
    val left = ((source.width - cropWidth) / 2).toInt()
    val top = ((source.height - cropHeight) / 2).toInt()

    // Ensure crop bounds are within the source bitmap
    val finalCropWidth = if (left < 0) source.width else cropWidth.toInt().coerceAtMost(source.width - left)
    val finalCropHeight = if (top < 0) source.height else cropHeight.toInt().coerceAtMost(source.height - top)

    // Center-crop the bitmap
    val croppedBitmap = Bitmap.createBitmap(source, left, top, finalCropWidth, finalCropHeight)

    // Scale the cropped bitmap back to desired size
    val scaledBitmap = Bitmap.createScaledBitmap(croppedBitmap, width.toInt(), height.toInt(), true)

    // Create a bitmap with transparency
    val output = Bitmap.createBitmap(width.toInt(), height.toInt(), Bitmap.Config.ARGB_8888)
    val canvas = Canvas(output)

    val paint = Paint().apply {
      isAntiAlias = true
      isFilterBitmap = true
      isDither = true
    }

    // Clear the canvas with a transparent color
    canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)

    // Draw the oval path
    val rect = RectF(0f, 0f, width.toFloat(), height.toFloat())
    canvas.drawOval(rect, paint)

    // Set the blending mode to SRC_IN to keep only the oval area
    paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
    canvas.drawBitmap(scaledBitmap, 0f, 0f, paint)

    return output
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

File: ios/Classes/ImageShapeCropperPlugin.swift
import Flutter
import UIKit

public class ImageShapeCropperPlugin: NSObject, FlutterPlugin {
    // Declare the channel as an instance variable
    private var channel: FlutterMethodChannel?

    // Register the plugin with the Flutter registrar
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "image_shape_cropper", binaryMessenger: registrar.messenger())
        let instance = ImageShapeCropperPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    // Handle incoming method calls from Flutter
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "cropOval":
            guard let args = call.arguments as? [String: Any],
                  let sourcePath = args["sourcePath"] as? String else {
                print("cropOval called with invalid arguments.")
                result(FlutterError(code: "INVALID_ARGUMENT", message: "sourcePath is missing", details: nil))
                return
            }

            let angle = args["angle"] as? Double ?? 0.0
            let width = args["width"] as? Double
            let height = args["height"] as? Double
            let scale = args["scale"] as? Double ?? 1.0
            let compressFormat = args["compressFormat"] as? String ?? "png"
            let compressQuality = args["compressQuality"] as? Int ?? 100

            print("cropOval called with sourcePath: \(sourcePath), angle: \(angle), width: \(String(describing: width)), height: \(String(describing: height)), scale: \(scale), compressFormat: \(compressFormat), compressQuality: \(compressQuality)")

            guard let image = UIImage(contentsOfFile: sourcePath) else {
                print("Cannot load image from path: \(sourcePath)")
                result(FlutterError(code: "INVALID_IMAGE", message: "Cannot load image", details: nil))
                return
            }

            // Rotate the image
            guard let rotatedImage = rotateImage(image: image, angle: CGFloat(angle)) else {
                print("Failed to rotate image.")
                result(FlutterError(code: "ROTATE_FAILED", message: "Failed to rotate image", details: nil))
                return
            }
            print("Image rotated by \(angle) degrees.")

            // Crop to oval with scaling
            let ovalImage = cropOvalImage(image: rotatedImage, width: width, height: height, scale: scale)
            print("Image cropped to oval.")

            // Compress the image
            var imageData: Data?
            if compressFormat.lowercased() == "png" {
                imageData = ovalImage.pngData()
                print("Compressing image to PNG format.")
            } else {
                let quality = CGFloat(compressQuality) / 100.0
                imageData = ovalImage.jpegData(compressionQuality: quality)
                print("Compressing image to JPEG format with quality \(compressQuality).")
            }

            if let data = imageData {
                print("Image compressed. Size: \(data.count) bytes.")
                result(data)
            } else {
                print("Failed to compress image.")
                result(FlutterError(code: "CROP_FAILED", message: "Failed to get image data", details: nil))
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // Helper method to rotate the image by a given angle
    private func rotateImage(image: UIImage, angle: CGFloat) -> UIImage? {
        let radians = angle * CGFloat.pi / 180
        var newSize = CGRect(origin: CGPoint.zero, size: image.size).applying(CGAffineTransform(rotationAngle: radians)).size
        newSize.width = floor(newSize.width)
        newSize.height = floor(newSize.height)

        UIGraphicsBeginImageContextWithOptions(newSize, false, image.scale)
        guard let context = UIGraphicsGetCurrentContext() else { return nil }

        // Move origin to middle
        context.translateBy(x: newSize.width / 2, y: newSize.height / 2)
        // Rotate around middle
        context.rotate(by: radians)
        // Draw the image at its center
        image.draw(in: CGRect(x: -image.size.width / 2, y: -image.size.height / 2,
                              width: image.size.width, height: image.size.height))

        let rotatedImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return rotatedImage
    }

    // Helper method to crop the image into an oval shape with scaling
    private func cropOvalImage(image: UIImage, width: Double?, height: Double?, scale: Double) -> UIImage {
        let width = width ?? Double(image.size.width)
        let height = height ?? Double(image.size.height)

        // Apply effectiveScale to clamp the scale value between 1.0 and 5.0
        let effectiveScale: Double
        if scale < 1.0 {
            effectiveScale = 1.0
        } else if scale > 5.0 {
            effectiveScale = 5.0
        } else {
            effectiveScale = scale
        }

        // Calculate the size of the crop rectangle based on scale
        let cropWidth = CGFloat(width) / CGFloat(effectiveScale)
        let cropHeight = CGFloat(height) / CGFloat(effectiveScale)

        // Calculate the top-left coordinates for center-cropping
        let left = (image.size.width - cropWidth) / 2
        let top = (image.size.height - cropHeight) / 2

        // Ensure crop bounds are within the source image
        let finalCropWidth = left < 0 ? image.size.width : min(cropWidth, image.size.width - left)
        let finalCropHeight = top < 0 ? image.size.height : min(cropHeight, image.size.height - top)

        // Center-crop the image
        guard let cgImage = image.cgImage?.cropping(to: CGRect(x: left, y: top, width: finalCropWidth, height: finalCropHeight)) else {
            return image
        }
        let croppedImage = UIImage(cgImage: cgImage, scale: image.scale, orientation: image.imageOrientation)

        // Scale the cropped image back to desired size with scale factor 1.0
        let scaledImage = UIGraphicsImageRenderer(size: CGSize(width: width, height: height), format: {
            let format = UIGraphicsImageRendererFormat.default()
            format.scale = 1.0 // Set scale to 1.0 to prevent automatic scaling
            return format
        }()).image { _ in
            croppedImage.draw(in: CGRect(x: 0, y: 0, width: CGFloat(width), height: CGFloat(height)))
        }

        // Create a new image with oval mask with scale factor 1.0
        let renderer = UIGraphicsImageRenderer(size: CGSize(width: width, height: height), format: {
            let format = UIGraphicsImageRendererFormat.default()
            format.scale = 1.0 // Set scale to 1.0 to prevent automatic scaling
            return format
        }())
        let ovalImage = renderer.image { context in
            // Draw the oval path
            let rect = CGRect(x: 0, y: 0, width: CGFloat(width), height: CGFloat(height))
            context.cgContext.addEllipse(in: rect)
            context.cgContext.clip()

            // Draw the scaled image within the oval
            scaledImage.draw(in: rect)
        }

        return ovalImage
    }

}
